packages/block-library/src/pattern/recursion-detector.js

```diff
@@ -0,0 +1,103 @@
+/**
+ * THIS MODULE IS INTENTIONALLY KEPT WITHIN THE PATTERN BLOCK'S SOURCE.
+ *
+ * This is because this approach for preventing infinite loops due to
+ * recursively rendering blocks is specific to the way that the `core/pattern`
+ * block behaves in the editor. Any other block types that deal with recursion
+ * SHOULD USE THE STANDARD METHOD for avoiding loops:
+ *
+ * @see https://github.com/WordPress/gutenberg/pull/31455
+ * @see packages/block-editor/src/components/recursion-provider/README.md
+ */
+
+/**
+ * @type {Map<string, Set<string>>}
+ */
+const patternDependencies = new Map();
```

Comment:
This is a global variable, so if you have two different block editors rendered on the same page, they'll share the same variable, do you think we should tie this to the "block-editor" store or something?

Reply:
The assumption I made to justify such a global was that patterns are uniquely named. So it shouldn't matter if there are more than one editor using different patterns.

I think it's good to take a step back and think about what we are guarding against. To break the assumption would mean that someone is somehow:

* loading two block editors simultaneously, **and**
* registering different patterns with those block editors, **and**
* creating naming conflicts between those patterns

That said, I'm open to the idea that, in principle, the lifetime of `patternDependencies` should be tied to the corresponding pattern registry. I like the sound of your suggestion to use registries as keys in a weak map.